### PR TITLE
fix(server): attach any workspace file to a chat message, and bound the read (#2029)

### LINKED ISSUE #2029: chat: attachments are rejected as "not a file in this company's workspace" — upload/list and resolve_attachments disagree on scope
## Symptom

Attaching a workspace file to a chat message fails. The POST is refused with:

> `attachment <node_id> is not a file in this company's workspace`

Reproduced during a live verification pass against a real host on current main, with **two independently-obtained nodes**:

1. a file **freshly uploaded** through the console's own upload endpoint, using the node id that endpoint returned;
2. a file **seeded into the company workspace** and confirmed present.

Both were rejected. So the endpoints that accept and list a file, and the resolver that consumes its id, do not agree on what counts as a workspace file.

## Where

`resolve_attachments`, `src/server/operator.rs:2490` — the rejection is at `:2523`. The predicate is:

```rust
let node = tree
    .iter()
    .find(|n| &n.id == node_id && n.is_binary())
```

Two candidate causes, both worth checking before fixing:

- **`is_binary()` is `mime.is_some()`.** A node the workspace happily stores and lists, but for which no mime was recorded, can never resolve as an attachment — it is not "not a file", it is a file with no mime. The error message asserts the former and hides the latter.
- **`runtime.workspace().tree(id)` scope.** If the tree read here is narrower than what upload/list operate on (different root, different company scoping, or staleness right after an upload), a genuinely present node is simply absent from this iteration.

## Why it matters

Attachments are unusable in chat for at least these two paths, and the failure is stated as if the operator supplied a bad id — so the message actively misdirects whoever hits it. It also means the attachment path is effectively unverified end to end: this was found only because a verification pass tried to exercise it and could not.

## Expected

A file that the upload endpoint accepted, or that the workspace lists, resolves as an attachment. Where it genuinely cannot (no mime, wrong kind), the error names the real reason rather than claiming the file is not in the workspace.

## Notes

Found during the union UI verification of #2020 / #1848 / #2005. **None of those three touch `src/server/operator.rs`** — this is pre-existing on main, not a regression from any of them.

### PR DESCRIPTION
**Pull request by oxoxDev**
## Summary

Attaching a workspace file to a chat message failed with `attachment <id> is not a file in this company's workspace` — a sentence wrong on both clauses.

`resolve_attachments` matched `n.is_binary()`, which is exactly `mime.is_some()`. But the workspace **deliberately** stores a text-typed, valid-UTF-8 file as a prose note with `mime: None`, and so does every seeded note, agent-authored note and console-created note. Those are files, they are in the company's tree, and `GET …/workspace/file/{id}` serves them. Only chat-upload sets a mime unconditionally — and that is the sole path every existing attachment test used, which is why CI was green while a human hit this on the first two files they tried.

So it was never an edge case about one node missing a mime: **every text file in the workspace was unattachable.**

Four-cell reproduction, run before any fix:

| upload | file | stored `mime` | attach |
|---|---|---|---|
| `…/workspace/upload` | `.md` | `null` | **400** |
| `…/workspace/upload` | `.png` | `image/png` | 200 |
| `…/chat/upload` | `.md` | `text/markdown` | 200 |
| `…/chat/upload` | `.png` | `image/png` | 200 |

Only the prose cell fails, so there is no second mechanism. The staleness/scoping alternative was searched and ruled out: every decorator forwards `tree()` verbatim, Mongo returns full documents with no secondary-read preference, and `upload()` already does a read-after-write.

Closes #2029

## API Or Behavior Changes

- `resolve_attachments` accepts any `NodeKind::File`; a folder is refused naming itself (`is a folder, not a file`) and an unknown id keeps its own message. The attachment cap, the dedup, `tree(id)` scoping and refuse-before-journal are untouched — they are deliberate hardening from a prior security review.
- For prose, `mime` comes from `mime_guess` on the **store's** name (never client-claimed) and `size` from the store's own measurement.
- `extracted_note_text` gives the model the note's words under the same two caps as binary extraction — without it a `.md` attaches and arrives empty.
- `read_blob` falls back to the text read for a prose node, served through the existing `serving_for(None)` — octet-stream, `attachment`, `nosniff`, never inline — with `Content-Length` set and no `ETag`. Without it the chip renders and its download 404s.
- **New port method `read_capped(company, id, max_bytes) -> Option<(node, body, len)>`**, returning the body's true byte length and the body only while it fits.

### Why the port grew a method

The original fix read a whole note into a `String` and only then compared its length to the 4 MiB cap — the ceiling applied *after* the allocation it exists to prevent, while the binary path it mirrors decides on `node.size` metadata and never opens an oversized file at all. A prose node carries no `size`, so **only the store can answer that cheaply**; a caller-side check is not expressible. Both halves come from one query, so a body cannot grow between the measurement and the read — this closes a read-after-stat race as well as the allocation.

Per backend, in its own currency: `fs` takes the length from the directory entry; sqlite from `length(CAST(content AS BLOB))` with the body admitted by a `CASE` in the same statement; mongodb from a `$strLenBytes` projection replacing an over-cap `content` server-side — `$strLenBytes` rather than `$strLenCP`, since the cap is bytes and code points would admit up to 4× it.

**Required, not defaulted.** A default would have let the three runtime decorators silently fall back to the unbounded `read` and lose the ceiling. The compiler named all 11 implementors instead: three real backends implement it, three decorators forward it, and the in-memory doubles use `read_capped_by_reading`, whose name is the warning that it is not a bound.

Two honest notes. `size` for a prose node is the store's measurement at resolve time rather than a value stamped beside a sha256 — nothing downstream compares them (`Attachment` carries no digest), but it is a genuine change of provenance. And the exposure this removes was **one note's body per resolve**, not twenty: the references are read sequentially and each body is dropped before the next. Still 16× the binary path's ceiling at the default 64 MiB blob cap, and now equal to it.

## Tests

Run unpiped, exit codes and counts captured:

- [x] `cargo fmt --all -- --check` · `cargo check --features openhuman` · `cargo clippy --locked --no-deps --features openhuman --all-targets -- -D warnings`
- [x] `cargo check --all-features --all-targets` — the real risk of a required trait method: no unimplemented implementor anywhere
- [x] `cargo test --lib --features openhuman` — **6779 passed** (was 6744)
- [x] `store::sqlite::` 57 passed · `store::mongodb::` 61 passed · attachment scope 21 passed
- [x] `npm run typecheck` / `:unit` / `:e2e`; `npx vitest run` — 4369 passed
- [x] `npx playwright test chat-attachment-workspace-file.spec.ts` — 5 passed

**The mongo lane actually ran** — rather than leave the aggregation compile-checked only, a throwaway `mongo:7` container served the full `store::mongodb::` suite including the new conformance test. `conformance::assert_workspace_read_capped` runs on all three backends and pins: bytes not characters, exact length for a withheld body, at-the-cap fits, empty distinguished from withheld, folders behaving as their `read` does, and company isolation.

Red proofs, verbatim. The original defect — all six new Rust tests fail on the unfixed source:

```
---- chat_message_attaches_a_note_uploaded_to_the_workspace stdout ----
assertion `left == right` failed: {"error":"invalid request: attachment … is not a file in this company's workspace"}
  left: 400   right: 200
```

And the bound:

```
---- an_over_cap_note_attachment_is_never_fully_read stdout ----
assertion `left == right` failed: resolving an over-cap attachment must not pull the note's body through the unbounded read
  left: 5242880   right: 0
```

That test wraps the permissive in-memory store in a byte-counting recorder placed **under** the runtime's decorators, so it measures what the request pulled through the whole production stack — a decorator that stops forwarding `read_capped` fails it the same way. Its sibling passed on both sides: the bound is not paid for by attachments that fit.

The Playwright spec was red-proofed too — reverting both source files and rebuilding gives 4 failed, 1 passed, each failure the same 400. The one that passes on both sides is the paperclip/PNG case, by design: it is the regression pin for the path that already worked.

**Verified in the running app**, pointed at a prompt-echoing inference fixture so the model's input was visible: the Engineer replied **quoting the note's own sentence**, its prompt carrying `[Attached file: … (text/markdown, 79 bytes) — workspace node …]` and the file's text. The chip's download returned 200, octet-stream, attachment, nosniff, no ETag, bytes identical. Repeated with a seeded note and a PNG.

Two things the e2e cannot drive, named rather than skipped: attaching an *existing* workspace node has no console affordance — the paperclip always uploads a fresh copy — so that step is posted over REST, exactly as the operator who filed this did; and `extracted_text` is deliberately absent from the console's wire shape, so the over-cap case's "carries no text" half is assertable only in the Rust suite.

### Review round: the `fs` backend's cap was still racy

`read_capped`'s single-call intent (above) held for sqlite and mongodb, which measure and read inside one query, but not for `fs`: it still did a `metadata()` stat and then a separate `read_optional()`, two calls with no shared snapshot. A concurrent `write` landing between them let the stat observe an under-cap old revision while the read materialized a larger replacement, so the returned body could exceed `max_bytes` while the reported length stayed stale. Verified against the code before fixing: `write_atomic` publishes by temp-file-plus-rename, and nothing serialises a `read_capped` call against it. Fixed by opening one file handle and doing both the length check and the read through it — a rename repoints the directory entry at a new inode rather than mutating the one already open, so the handle keeps referring to exactly the bytes it opened.

Also rebased onto current `upstream/main` (this branch was cut before it moved) and fixed the red "Markdown line cap" check by splitting `ports-console.md`'s `WorkspaceStore` section into its own page — see Documentation below.

Gates for this round, each run unpiped with its own exit code:

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --locked --no-deps --features openhuman --all-targets -- -D warnings`
- [x] `cargo test --locked --features openhuman --lib` — 6817 passed, 0 failed, 3 ignored
- [x] `bash scripts/ci/assert-md-line-cap.sh` — every file 500 lines or fewer

New regression test (`assert_workspace_read_capped_race`, shared across `fs`/sqlite/mongodb conformance): races readers against a writer alternating an under-cap and an over-cap revision, and asserts every returned body stays within the cap and its length matches the bytes actually returned. Verified against the previous two-call `fs` implementation first — fails reliably (`read_capped returned a 600000-byte body against a 300000-byte cap`) — then against the fix, where it passed on repeated runs.

Verified live against a locally built host (`--features openhuman,mcp`), inference pointed at staging: created a `.md` note containing a distinctive sentence, attached it to a chat message, and the agent's reply quoted the sentence back verbatim — the chip's reported name and size (`121` bytes) matched the note exactly, and downloading through the blob route returned the identical bytes. A 5 MiB over-cap note still attached, its chip still showed the correct name and true size (`5242880`), the send did not fail (200), and the agent's reply reported it had no readable text to quote rather than fabricating content. A folder id was refused with `attachment … is a folder, not a file`, and never reached the transcript. The full `chat-attachment-workspace-file.spec.ts` suite (5 tests) also re-passed against this round's build.

This round makes no change to what the attachment path returns on the wire — the fix is purely a concurrency correctness fix inside the store, not a shape or behavior change — so the committed Playwright spec did not need extending; the new Rust conformance test is the one that can actually drive the race deterministically enough to be a regression pin.

### Review round: oversized caps wrapped negative in sqlite and mongodb

`read_capped`'s cap is bound as `max_bytes as i64`. A caller passing a cap above `i64::MAX` (an "effectively unlimited" request) wrapped that cast negative, so both database backends' `<= cap` comparisons were false for every note regardless of size — an over-cap read that should have been unlimited instead withheld everything. `fs` was unaffected: it compares `len > max_bytes` in `u64` throughout. Fixed by saturating the cap to `i64::MAX` before binding in both sqlite and mongodb (mongodb had the identical cast in its `$lte` projection).

New regression test on the sqlite backend (`read_capped_clamps_oversized_max_bytes`): creates a small note, reads it with `max_bytes: u64::MAX`, and asserts the body comes back whole. Verified as a red proof against the unclamped cast first — fails with the body withheld (`left: "" right: "small body"`) — then passes against the fix; the full sqlite (59 tests) and mongodb (62 tests, against a real `mongod`) lanes were re-run afterward with no other regressions.

Also added `read_capped`'s signature to the `WorkspaceStore` trait excerpt in `ports-console-workspace.md` (added in the previous round's split but left off the excerpt), and rebased onto `upstream/main` again (moved several times since the previous round).

A third finding — the `read_blob` fallback for a prose note buffers the whole body via `read()` before `one_chunk` wraps it, so a sufficiently large note costs the request its full size in memory, and text notes are deliberately unmetered — is confirmed valid but left unaddressed here: the fix is a genuine streaming text read added to `WorkspaceStore` and implemented per backend (chunked file I/O for `fs`, incremental reads for sqlite/mongodb), which is a port-contract change across three backends rather than something this attachment-scope PR should carry. Flagged on the thread as follow-up work.

Gates for this round, each run unpiped with its own exit code:

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --locked --no-deps --features openhuman --all-targets -- -D warnings`
- [x] `cargo test --locked --features openhuman --lib` — 6866 passed, 0 failed, 3 ignored
- [x] `bash scripts/ci/assert-md-line-cap.sh` — every file 500 lines or fewer
- [x] `cargo test --locked --features sqlite --lib store::sqlite` — 59 passed (re-run post-rebase)
- [x] `cargo test --locked --features mongodb --lib store::mongodb` — 62 passed, against a real `mongod`

## Documentation

`docs/spec/runtime/api.md` and `docs/spec/runtime/ports-console.md` both stated the old rule; the latter now documents `read_capped` beside the text/bytes split, including why a decorator forwards it.

`ports-console.md` was 518 lines against the repository's 500-line cap once that documentation landed. Split `WorkspaceStore` — its largest section — into `docs/spec/runtime/ports-console-workspace.md`, left a pointer under the old heading (`ports.md`'s existing `#workspacestore` anchor still lands there), and indexed the new page from the runtime `README.md`.

The split page's trait excerpt listed every `WorkspaceStore` method except `read_capped` itself, despite the prose below it describing that method at length. Added its signature to the excerpt, in the same position it holds in `src/ports/workspace.rs` (right after `read`).

## Summary by CodeRabbit

- **New Features**
  - Workspace notes can now be attached to chat messages, alongside binary files.
  - Prose notes are available for download through workspace file links.
  - Oversized notes remain attachable while respecting extraction limits.
  - Added documentation covering workspace storage, note trees, folder claims, search, quotas, and system roots.

- **Bug Fixes**
  - Folders and missing workspace items are correctly rejected as attachments.
  - Note MIME types and byte sizes are determined accurately when served or attached.
  - Workspace reads now enforce byte limits consistently and prevent inconsistent results during concurrent updates.

### CODE REVIEW

File: src/store/fs_ops.rs [hunk 1]
```diff
@@ -1624,6 +1624,34 @@ impl WorkspaceStore for FsOps {
...
+        // costs a `stat` rather than a read of the file it names.
+        let len = match tokio::fs::metadata(&path).await {
+            Ok(meta) => meta.len(),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => 0,
+            Err(e) => return Err(crate::store::fs::io_err(&path, e)),
+        };
+        if len > max_bytes {
+            return Ok(Some((node, String::new(), len)));
+        }
+        let content = read_optional(&path).await?;
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/store/fs_ops.rs]: Confirmed: `write_atomic` publishes by rename, and nothing serialises this read against it, so the stat and the later open really could observe different revisions. Fixed in d01e09d63 — `read_capped` now opens one file handle and measures/reads from that same handle, so the length and the body always describe the same snapshot. Added a regression test (`assert_workspace_read_capped_race`, e05b70c91) that races a writer against concurrent capped reads and asserts a returned body never exceeds the cap and its length always matches the bytes returned; it reliably fails against the old two-call implementation and passes against the fix, and runs on all three backends. No other stat-then-open call sites in this module.

File: src/store/sqlite.rs [hunk 2]
```diff
@@ -3532,6 +3532,46 @@ impl crate::ports::workspace::WorkspaceStore for SqliteStore {
...
+        // One statement, so the length and the body it admits describe the same
+        // row: the cap is applied by SQLite, and an over-cap body is never
+        // carried across the boundary into a Rust `String`.
+        let row: Option<(String, i64, String)> = conn
+            .query_row(
+                "SELECT node_json, \
+                        length(CAST(content AS BLOB)), \
+                        CASE WHEN length(CAST(content AS BLOB)) <= ?3 THEN content ELSE '' END \
+                 FROM workspace_nodes WHERE company_id = ?1 AND id = ?2",
+                params![company.as_ref(), id, max_bytes as i64],
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/store/sqlite.rs]: Addressed in 30a776588 — clamped max_bytes to i64::MAX before binding in sqlite; mongodb had the identical cast (`max_bytes as i64` in the $lte projection) so I fixed it there too and added a regression test on the sqlite backend that fails against the unclamped cast (a cap of u64::MAX withheld a small body before the fix).

File: docs/spec/runtime/ports-console-workspace.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+    // Bytes (#553) — additive; the text path above is untouched.
+    async fn create_binary(&self, company: &CompanyId, node: &WorkspaceNode,
+                           bytes: &[u8]) -> Result<()>;
+    async fn write_binary(&self, company: &CompanyId, id: &str, bytes: &[u8],
+                          mime: Option<&str>, author: WorkspaceOrigin)
+        -> Result<WorkspaceNode>;
+    async fn read_bytes(&self, company: &CompanyId, id: &str)
+        -> Result<Option<(WorkspaceNode, BlobStream)>>;
+}
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply docs/spec/runtime/ports-console-workspace.md:40]: Addressed in 30a776588 — added read_capped's signature to the trait excerpt, in the same position it holds in src/ports/workspace.rs (right after read).

File: src/server/ops/workspace.rs [hunk 4]
```diff
@@ -556,15 +557,37 @@ async fn read_blob(
...
+            else {
+                return Err(missing());
+            };
+            if node.kind != NodeKind::File || node.is_binary() {
+                return Err(missing());
+            }
+            let bytes = content.into_bytes();
+            let size = Some(bytes.len() as u64);
+            (node, crate::ports::workspace::one_chunk(bytes), size)
+        }
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/server/ops/workspace.rs:590]: Confirmed valid, leaving unaddressed here rather than half-fixing. walk_workspace and write() both accept prose of any size — text is deliberately unmetered (max_blob_mb/tree_quota_gb only count binary payloads) — so a large note is reachable. But the fallback's memory cost is already paid by the read() call above it: the full body is materialized as a String, then into_bytes(), before one_chunk ever runs, so wrapping that single allocation in a multi-chunk stream wouldn't lower peak memory per request. A real fix needs a genuine streaming text read added to WorkspaceStore and implemented per backend (chunked file read for fs, incremental read for sqlite/mongodb) — a port-contract change across three backends, not something this PR's attachment-scope fix should carry. Proposing it as separate follow-up work rather than doing it partially here.

**Comment by oxoxDev** [comment]
@coderabbitai review

File: src/server/ops/workspace.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/server/ops/workspace.rs:590]: Holding the earlier position, with the reasoning restated since this came back.

The concern is valid and I am not disputing the mechanism: `read_blob`'s fallback calls unbounded `read()` and materializes the whole note before `one_chunk` wraps it, so chunking the response afterwards does not reduce peak memory — the allocation has already happened. Concurrent downloads of large prose notes can therefore hold one full note each.

What makes it out of scope here is that there is no bounded or streaming text read to switch to. `read_capped` bounds a read but still returns an owned body, so it changes the ceiling, not the shape. A real fix adds a streaming text read to the `WorkspaceStore` port and implements it across the fs, sqlite and mongodb backends — a port-contract change touching three implementations, none of which this PR otherwise goes near.

This PR bounds the *attachment* read, which is the path it exists to fix. Silently widening it into a port change would make it unreviewable against its own description. Tracked as follow-up work rather than dropped.
